README.md
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/autoplotly)](https://cran.r-project.org/package=autoplotly)

# autoplotly

This R package provides functionalities to automatically generate interactive visualizations for many
popular statistical results supported by [ggfortify](https://github.com/sinhrks/ggfortify)
package with [plotly.js](https://plot.ly) and [ggplot2](http://ggplot2.tidyverse.org/) style.
The generated visualizations can also be easily extended using ggplot2 syntax while staying interactive. Please check out the introduction [here](https://terrytangyuan.github.io/2018/02/12/autoplotly-intro/).

## Installation

To install the current version from CRAN, use:

``` r
install.packages("autoplotly")
```

To install from development version on Github, use:

``` r
devtools::install_github("terrytangyuan/autoplotly")
```

## Example

``` r
# Automatically generate interactive plot for results produced by `stats::prcomp`
p <- autoplotly(prcomp(iris[c(1, 2, 3, 4)]), data = iris,
  colour = 'Species', label = TRUE, label.size = 3, frame = TRUE)

# You can apply additional ggplot2 elements to the generated interactive plot
p +
  ggplot2::ggtitle("Principal Components Analysis") +
  ggplot2::labs(y = "Second Principal Components", x = "First Principal Components")

# Or apply additional plotly elements to the generated interactive plot
p %>% plotly::layout(annotations = list(
  text = "Example Text",
  font = list(
    family = "Courier New, monospace",
    size = 18,
    color = "black"),
  x = 0,
  y = 0,
  showarrow = TRUE))
```

You can `autoplotly` many other statistical results automatically with the help of [ggfortify](https://github.com/sinhrks/ggfortify). A complete list can be found [here](https://github.com/sinhrks/ggfortify#coverage).


NEWS
autoplotly dev
---------------------

* TBA

autoplotly v0.1.1
---------------------

* This is essentially the same as the v0.1.0. The previous version was archived due to one of the dependencies.

autoplotly v0.1.0
---------------------

* First release which includes the single `autoplotly()` function.


tests/testthat.R
library(testthat)
library(autoplotly)

test_check("autoplotly")


tests/testthat/test-all.R
context("Testing all functionalities")

test_that("autoplotly works correctly", {
  df <- iris[c(1, 2, 3, 4)]

  a <- autoplotly(prcomp(df), data = iris, colour = 'Species', label = TRUE, label.size = 3, frame = TRUE)

  # Composable with ggplot2
  b <- a + ggplot2::ggtitle("This is a custom title") + ggplot2::labs(y = "Second Principal Components", x = "First Principal Components")
})


tests/testthat/test_all.R
context("Testing all functionalities")

df <- iris[c(1, 2, 3, 4)]
result <- prcomp(df)

test_that("autoplotly works correctly", {
  p <- autoplotly(result, data = iris, colour = 'Species', label = TRUE, label.size = 3, frame = TRUE)
  expect_true(inherits(p, "plotly"))
  expect_true(inherits(p$ggplot_obj, "ggplot"))
})

test_that("autoplotly is composable with additional ggplot2 elements", {
  p <- autoplotly(result, data = iris, colour = 'Species', label = TRUE, label.size = 3, frame = TRUE)
  p <- p + ggplot2::ggtitle("This is a custom title") + ggplot2::labs(y = "Second Principal Components", x = "First Principal Components")
  expect_true(inherits(p, "plotly"))
  expect_true(inherits(p$ggplot_obj, "ggplot"))
})

test_that("Errors are thrown when class is not supported by ggfortify", {
  a <- 1
  class(a) <- "None"
  expect_error(autoplotly(a, data = iris, colour = 'Species', label = TRUE, label.size = 3, frame = TRUE), "Only object of the following classes are supported for autoplotly")
})

test_that("autoplotly accepts additional plotly args", {
  p <- autoplotly(result, data = iris, colour = 'Species', label = TRUE, label.size = 3, frame = TRUE,
                  tooltip = "all", dynamicTicks = FALSE, layerData = 1, originalData = TRUE, source = "A")
  expect_true(inherits(p, "plotly"))
  expect_true(inherits(p$ggplot_obj, "ggplot"))
})


R/autoplotly.R
#' @export
autoplotly.default <- function(
  object,
  ...,
  width = NULL,
  height = NULL,
  tooltip = "all",
  dynamicTicks = FALSE,
  layerData = 1,
  originalData = TRUE,
  source = "A",
  margin = 0.02,
  shareX = TRUE,
  shareY = TRUE,
  titleX = shareX,
  titleY = shareY,
  which_layout = "merge") {

  autoplot_methods <- gsub("autoplot.", "", grep("autoplot.", ls(getNamespace("ggfortify")), value = TRUE))
  if (!any(class(object) %in% autoplot_methods)) {
    stop(paste0(
      "Only object of the following classes are supported for autoplotly: ",
      paste0(autoplot_methods, collapse = ", ")))
  }
  ggplot_obj <- autoplot(object, ...)
  if (is(ggplot_obj, "ggmultiplot")) {
    nplots <- length(ggplot_obj@plots)
    layout <- get.layout(nplots, ggplot_obj@ncol, ggplot_obj@nrow)
    plotly_obj <- subplot(
      lapply(ggplot_obj@plots, function(p) p + ggplot2::ggtitle("")),
      shareX = shareX,
      shareY = shareX,
      titleX = titleX,
      titleY = titleY,
      margin = margin,
      # ncols = ncol(layout), # TODO: plots[[i]][["frames"]] : subscript out of bounds. Potentially plotly issue
      # nrows = nrow(layout),
      which_layout = which_layout)
  } else {
    plotly_obj <- ggplotly(
      ggplot_obj,
      width = width,
      height = height,
      tooltip = tooltip,
      dynamicTicks = dynamicTicks,
      layerData = layerData,
      originalData = originalData, source = source)
  }
  plotly_obj$ggplot_obj <- ggplot_obj
  plotly_obj
}

get.layout <- function(nplots, ncol, nrow) {
  if (ncol == 0 && nrow == 0) {
    ncol <- 2
  } else if (ncol == 0 && nrow != 0) {
    ncol <- ceiling(nplots / nrow)
  }
  if (nrow == 0) {
    nrow <- ceiling(nplots / ncol)
  } else {
    nrow <- nrow
  }
  if (nrow * ncol < nplots) {
    message <- paste('nrow * ncol (', nrow, '*', ncol,
                     ')must be larger than number of plots', nplots)
    stop(message)
  }
  t(matrix(1:(ncol * nrow), ncol = nrow, nrow = ncol))
}


R/generic_methods.R
#' @export
"+.plotly" <- function(p, ggplot_element) {
  ggplot_obj <- tryCatch(
    p$ggplot_obj + ggplot_element,
    # Slightly better error message
    error = function(e) {
      if (grepl("Don't know how to add", e$message)) {
        stop("You are using ggplot2 with incorrect syntax.",
             "Please visit http://ggplot2.tidyverse.org/index.html for tutorials.")
      }
    }
  )
  autoplotly(ggplot_obj)
}

#' Automatic Visualization of Popular Statistical Results Using 'plotly.js' and 'ggplot2'
#'
#' @description This function provides functionality to automatically generate interactive
#' plot for many popular statistical results supported by 'ggfortify' package using 'plotly.js' and 'ggplot2'.
#'
#' @param object The object that represents your statistical result, e.g. `stats::prcomp(iris[-5])`.
#' @param ... Arguments passed to `ggfortify::autoplot` function for the applied
#'   object. For example, if your object is constructed from `stats::prcomp(iris[-5])`,
#'   you can find the documentation for the list of additional arguments via
#'   `?ggfortify:::autoplot.prcomp()`.
#' @inheritParams plotly::ggplotly
#' @inheritParams plotly::subplot
#' @export
#' 
#' @examples
#' # Automatically generate interactive plot for results produced by `stats::prcomp`
#' p <- autoplotly(prcomp(iris[c(1, 2, 3, 4)]), data = iris,
#'                 colour = 'Species', label = TRUE, label.size = 3, frame = TRUE)
#' 
#' # You can apply additional ggplot2 elements to interactive plot built using `autoplotly()`
#' p +
#'   ggplot2::ggtitle("Principal Components Analysis") +
#'   ggplot2::labs(y = "Second Principal Components", x = "First Principal Components")
#' 
#' # Or apply additional plotly elements to the generated interactive plot
#' p %>% plotly::layout(annotations = list(
#'   text = "Example Text",
#'   font = list(
#'     family = "Courier New, monospace",
#'     size = 18,
#'     color = "black"),
#'   x = 0,
#'   y = 0,
#'   showarrow = TRUE))
autoplotly <- function(object,
                       ...,
                       width = NULL,
                       height = NULL,
                       tooltip = "all",
                       dynamicTicks = FALSE,
                       layerData = 1,
                       originalData = TRUE,
                       source = "A",
                       widths = NULL,
                       heights = NULL,
                       margin = 0.02,
                       shareX = TRUE,
                       shareY = TRUE,
                       titleX = shareX,
                       titleY = shareY,
                       which_layout = "merge") {
  UseMethod("autoplotly")
}
